src/partners/thorchain.ts

```diff
@@ -52,169 +61,247 @@ const asThorchainResult = asObject({
-        in: txIns,
-        out: txOuts,
-        pools,
-        status: txStatus
-      } = asThorchainTx(rawTx) // Check RAW trasaction
-
-      const status = 'complete'
-      if (txStatus !== 'success') {
-        continue
+    let offset = 0
+
+    let headers: HeadersInit | undefined
+    if (xClientId != null) {
+      headers = {
+        'x-client-id': xClientId
       }
-      if (pools.length !== 2) {
-        continue
+    }
+
+    while (!done) {
+      const url = `https://${midgardUrl}/v2/actions?address=${thorchainAddress}&type=swap,refund&affiliate=${affiliateAddress}&offset=${offset}&limit=${LIMIT}`
+      let jsonObj: ThorchainResult
+      try {
+        await snooze(500)
+        const result = await retryFetch(url, {
+          method: 'GET',
+          headers
+        })
+        if (!result.ok) {
+          const text = await result.text()
+          throw new Error(`Thorchain error: ${text}`)
+        }
+        const resultJson = await result.json()
+        jsonObj = asThorchainResult(resultJson)
+      } catch (e) {
+        datelog(e)
+        throw e
       }
+      const txs = jsonObj.actions
+      for (const rawTx of txs) {
+        const {
+          date,
+          metadata,
+          in: txIns,
+          out: txOuts,
+          pools,
+          status: txStatus
+        } = asThorchainTx(rawTx) // Check RAW trasaction
 
-      const srcAsset = txIns[0].coins[0].asset
-      const match = txOuts.find(o => {
-        const match2 = o.coins.find(c => c.asset === srcAsset)
-        return match2 != null
-      })
+        const { swap } = metadata
+        if (swap?.affiliateAddress !== affiliateAddress) {
+          continue
+        }
+        const status = 'complete'
+        if (txStatus !== 'success') {
+          continue
+        }
 
-      // If there is a match between source and dest asset that means a refund was made
-      // and the transaction failed
-      if (match != null) {
-        continue
-      }
+        // There must be an affiliate output
+        const affiliateOut = txOuts.some(
+          o => o.affiliate === true || o.address === thorchainAddress
+        )
+        if (!affiliateOut) {
+          continue
+        }
+
+        // Find the source asset
+        if (txIns.length !== 1) {
+          throw new Error(
+            `${pluginId}: Unexpected ${txIns.length} txIns. Expected 1`
+          )
+        }
+        const txIn = txIns[0]
+        if (txIn.coins.length !== 1) {
+          throw new Error(
+            `${pluginId}: Unexpected ${txIn.coins.length} txIn.coins. Expected 1`
+          )
+        }
+        const coin = txIn.coins[0]
+        const srcAsset = coin.asset
+        const depositAmount = Number(coin.amount) / THORCHAIN_MULTIPLIER
 
-      const timestampMs = div(date, '1000000', 16)
-      const { timestamp, isoDate } = smartIsoDateFromTimestamp(
-        Number(timestampMs)
-      )
+        if (srcAsset == null || depositAmount == null) {
```

Comment:
Why the null checks if these types cannot be null?

Reply:
ah. leftover from prior code structure where they could be null